src/v/serde/avro/parser.cc

```diff
@@ -0,0 +1,180 @@
+#include "serde/avro/parser.h"
+
+#include "bytes/iobuf_parser.h"
+
+namespace serde::avro {
+static constexpr int max_nested_depth = 200;
+namespace {
+
+struct parser_state {
+    int level{0};
+};
+
+class parser {
+public:
+    parser(iobuf buf, const ::avro::ValidSchema& schema)
+      : _parser(std::move(buf))
+      , _schema(&schema) {}
+
+    ss::future<std::unique_ptr<parsed::message>>
+    parse_node(const ::avro::NodePtr& node, parser_state& state) {
+        state.level++;
+        if (state.level >= max_nested_depth) {
+            throw std::invalid_argument(fmt::format(
+              "max nested field depth of {} reached", max_nested_depth));
+        }
+
+        switch (node->type()) {
+        case ::avro::AVRO_STRING: {
+            auto sz = parse_length();
+            state.level--;
+            co_return std::make_unique<parsed::message>(_parser.share(sz));
+        }
+        case ::avro::AVRO_BYTES: {
+            auto sz = parse_length();
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              parsed::primitive(_parser.share(sz)));
+        }
+        case ::avro::AVRO_INT: {
+            auto [value, _] = _parser.read_varlong();
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              parsed::primitive(static_cast<int32_t>(value)));
+        }
+        case ::avro::AVRO_ENUM:
+        case ::avro::AVRO_LONG: {
+            auto [value, _] = _parser.read_varlong();
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              parsed::primitive(value));
+        }
+        case ::avro::AVRO_FLOAT:
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              *reinterpret_cast<float*>(_parser.read_bytes(4).begin()));
+        case ::avro::AVRO_DOUBLE:
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              *reinterpret_cast<double*>(_parser.read_bytes(8).begin()));
```

Comment:
You can replace this `reienterpret_cast` with `consume_type` here and other places